tests/components/vizio/test_media_player.py

```diff
@@ -0,0 +1,289 @@
+    """Test for Vizio Speaker entity when on."""
+    await _test_vizio_init(
+        hass,
+        MOCK_SPEAKER_CONFIG,
+        VIZIO_DEVICE_CLASS_SPEAKER,
+        DEVICE_CLASS_SPEAKER,
+        True,
+        STATE_ON,
+    )
+
+
+async def test_vizio_speaker_off(hass: HomeAssistantType) -> None:
+    """Test for Vizio Speaker entity when off."""
+    await _test_vizio_init(
+        hass,
+        MOCK_SPEAKER_CONFIG,
+        VIZIO_DEVICE_CLASS_SPEAKER,
+        DEVICE_CLASS_SPEAKER,
+        False,
+        STATE_OFF,
+    )
+
+
+async def test_vizio_speaker_unavailable(hass: HomeAssistantType) -> None:
+    """Test for Vizio Speaker entity when unavailable."""
+    await _test_vizio_init(
+        hass,
+        MOCK_SPEAKER_CONFIG,
+        VIZIO_DEVICE_CLASS_SPEAKER,
+        DEVICE_CLASS_SPEAKER,
+        None,
+        STATE_UNAVAILABLE,
+    )
+
+
+async def test_vizio_init_tv_on(hass: HomeAssistantType) -> None:
+    """Test for Vizio TV entity when on."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        True,
+        STATE_ON,
+    )
+
+
+async def test_vizio_init_tv_off(hass: HomeAssistantType) -> None:
+    """Test for Vizio TV entity when off."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        False,
+        STATE_OFF,
+    )
+
+
+async def test_vizio_init_tv_unavailable(hass: HomeAssistantType) -> None:
+    """Test for Vizio TV entity when unavailable."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        None,
+        STATE_UNAVAILABLE,
+    )
+
+
+async def test_setup_failure(hass: HomeAssistantType) -> None:
+    """Test media player entity setup failure."""
+    with patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.can_connect",
+        return_value=False,
+    ):
+        try:
+            await async_setup_entry(hass, MOCK_SPEAKER_CONFIG_ENTRY, None)
+        except Exception as e:
+            assert isinstance(e, PlatformNotReady)
+
+        try:
+            await async_setup_entry(hass, MOCK_TV_CONFIG_ENTRY, None)
+        except Exception as e:
+            assert isinstance(e, PlatformNotReady)
+
+
+async def test_services(hass: HomeAssistantType) -> None:
+    """Test media player entity services."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        True,
+        STATE_ON,
+    )
+
+    with patch("homeassistant.components.vizio.media_player.VizioAsync.pow_on"):
+        await hass.services.async_call(
+            MP_DOMAIN,
+            SERVICE_TURN_ON,
+            service_data={ATTR_ENTITY_ID: ENTITY_ID},
+            blocking=True,
+        )
+
```

Comment:
After each of these service calls we should assert that the corresponding library method was called.